### ISSUE TITLE
Allow users to control error/warning handling (like numpy.seterr)

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#727]
## Feature Request

**Is your feature request related to a problem? Please describe.**
When developing and debugging, it may be desirable that an uncaught exception is raised.  This allows the developer to use the debugger to inspect the stack at this point.  The current Satpy behaviour, to log the problem and try to proceed as well as possible, is desirable for operational usage, but complicates debugging and may make some bugs go unnoticed

**Describe the solution you'd like**

I would like a system like `numpy.seterr` — perhaps even *using* `numpy.seterr` — such that the user can control what to do if things go wrong: ignore, log, warn, error, or function call.

**Describe any changes to existing user workflow**
The current behaviour would remain the default, so this change should be 100% backward compatible.

**Additional context**
I've found myself replacing or commenting out `Except:`-statements in source, which is a rather cumbersome approach.

**Comment by djhoese** [comment]
My preferred solution would be to make this part of the donfig config object that I hope to add this year (#628). It could be used as a context manager and would only have to be used in a couple key places where exceptions are ignored (yaml_reader.py when failing to load a dataset, etc).

**Comment by gerritholl** [comment]
On a related note, one could discuss where Satpy and other Pytroll packages use `logging.warning`, and in what cases `warnings.warn` would be more appropriate.  From the [logging HOWTO](https://docs.python.org/3.7/howto/logging.html):

> `warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning

> `logging.warning()` if there is nothing the client application can do about the situation, but the event should still be noted

Arguably, when Satpy is warning `Don't know how to open the following files:`, that is an avoidable issue where the client application should be modified, so (if Python guidelines are to follow) it should use `warnings.warn()`, rather than `logging.warning()` as it's doing now.  If Pytroll has good reasons to explicitly have a different policy, that should probably be noted in the [Pytroll coding guidelines](http://pytroll.github.io/guidelines.html).

**Comment by djhoese** [comment]
Good point. Do you have a list of logged warnings that you think should be warnings.warn warnings?

For the not sure how to open the files thing, that code has changed a lot in the last couple years and I think we did have a lot of discussion about how the user should be notified about it. In the end it seemed more likely to us that users would provide all the files in a directory and try to read what they could. If Satpy didn't know how to read some of the files then we didn't want to make a big fuss about it...but you're right, maybe we should.

**Comment by gerritholl** [comment]
I just discovered that there exists a `logging_on` utility that at least allows a bit more control than the `debug_on` utility.

**Comment by djhoese** [comment]
You could use the basic logging interfaces and control the logging level yourself:

```
import logging
logging.getLogger('satpy').setLevel(logging.INFO)
logging.getLogger('pyresample').setLevel(logging.ERROR)
```

I think that should work as far as levels go and doesn't effect other loggers like logging_on does.

**Comment by gerritholl** [comment]
An example use case: I'm loading datasets, then resampling, then calculating a product from the result.  If some of the datasets are not loadable further processing is pointless.  It would be useful to either have a global setting or a flag I can pass to `.load(..., strict=True)`, such that the logged message 

https://github.com/pytroll/satpy/blob/bfc7bfb58c7c3b8b6472e6babad0aba1da6a75ac/satpy/readers/yaml_reader.py#L770-L771

would be an exception instead.  As it is now, I'd need to capture and parse log messages and if reading fails due to a known cause, have some kind of handling, and if it fails due an unknown cause, raise an exception.

**Comment by gerritholl** [comment]
In theory I could create a custom logging `Handler` that raises an exception when it gets something logged with the `error` level, although I have a feeling that would be a waste of time as I believe exceptions raised in logging handlers are swallowed in general.

It seems different parts of Satpy have different ideas about what should be warnings and what should be errors, and `logger.error` is rarely used.  For example:

https://github.com/pytroll/satpy/blob/bfc7bfb58c7c3b8b6472e6babad0aba1da6a75ac/satpy/readers/fci_l2_nc.py#L100

is a lot less serious than

https://github.com/pytroll/satpy/blob/bfc7bfb58c7c3b8b6472e6babad0aba1da6a75ac/satpy/readers/yaml_reader.py#L691-L693

and certainly than

https://github.com/pytroll/satpy/blob/bfc7bfb58c7c3b8b6472e6babad0aba1da6a75ac/satpy/writers/scmi.py#L1126-L1129

which, judging from the code, doesn't appear to be true.

I think we may need some guidelines in the developer guide on when to use error, warning, info, debug, and trace, respectively, in particular if we're going to have better programmatic control.  Ideally I'd like to be able to set that anything of severity x and above should be an exception and anything below should be ignored, but then it's important that code doesn't issue a warning for something that doesn't actually matter (the scmi reader should IMHO be debug level).

**Comment by djhoese** [comment]
> which, judging from the code, doesn't appear to be true.

Why do you say that?

> I think we may need some guidelines in the developer guide on when to use error, warning, info, debug, and trace, respectively

Sounds good to me.

**Comment by gerritholl** [comment]
> > which, judging from the code, doesn't appear to be true.
> 
> Why do you say that?

Because the line is immediately followed by a call to `save_datasets`, so the only difference in performance is the overhead of a single function call, which is tiny compared to all the work on resampling and I/O.  I don't see how there can be any significant difference in performance between `save_dataset` and `save_datasets` if the former immediately calls the latter (in fact, the biggest difference in performance is due to the log message…).

**Comment by gerritholl** [comment]
One way to implement configurable behaviour (log vs. raise) would be to have a helper function somewhere in pytroll (probably in satpy).  Instead of directly calling `logger.warning(msg)`, the code would call `satpy.utils.warning(logger, msg)`, which would have configurable behaviour (log vs. raise vs. callback).  This would require replacing all logger calls throughout pytroll , which is quite significant, but there's probably no way around it.

**Comment by gerritholl** [comment]
Turns out my ugly hack on intentionally raising an exception in a logger actually works (see linked PR in my own repo), but I suspect we don't want to go that way in pytroll because it's not what logging handlers are meant for.

**Comment by djhoese** [comment]
As for scmi performance, it isn't the performance of that individual save. It is the performance of saving multiple variables where save_datasets will perform vetter when given all the data that will be saved. It will combine calculations that it can to perform better.

**Comment by gerritholl** [comment]
OK, I see what the intention is now.  I still find it a strange message.  Suppose I want to save a single dataset, I use a method that saves a single dataset, and then I get a message that calling `save_datasets` has better performance.  There is nowhere in the message that says that this warning applies if you save multiple datasets and the message still appears if you save only one.

**Comment by gerritholl** [comment]
@djhoese Now that there is a central configuration object (since nearly 6 months), what are your thoughts on a system to control warning messages as exceptions yes or no?

**Comment by djhoese** [comment]
Sounds good as long as defaults are the same behavior.
